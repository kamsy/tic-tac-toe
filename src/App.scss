.App {
  text-align: center;
  display: flex;
  height: calc(100vh - 100px);
  justify-content: center;
  padding-top: 100px;

  button {
    outline: none;
    box-shadow: none;
    border-radius: 4px;
    padding: 12px 16px;
    border: none;
    cursor: pointer;

    &:disabled {
      opacity: 0.3;
      cursor: not-allowed;
    }

    &:not(.reset) {
      background: rgb(0, 188, 100);
      color: #ffffff;

      &:hover {
        background: rgb(0, 209, 122);
      }
    }

    &.reset {
      background-color: rgb(230, 230, 230);
      color: rgb(0, 0, 0);
    }
  }

  .winner {
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
  }

  .game {
    display: flex;
    flex-direction: column;
    align-items: center;
    margin-right: 40px;
    margin-left: 400px;

    .turn__indicator {
      text-align: left;
      width: 100%;
    }

    .grid {
      display: grid;
      transition: 0.25s all ease-in-out;

      &.disable {
        opacity: 0.3;
        cursor: not-allowed;
      }

      &-cell {
        transition: 0.25s all ease-in-out;
        border: 1px solid grey;
        width: 56px;
        height: 56px;
        display: flex;
        align-items: center;
        justify-content: center;
        position: relative;

        span {
          opacity: 0;
        }

        &.disable {
          cursor: not-allowed;
        }

        &::before,
        &::after {
          content: "";
          position: absolute;
          top: 50%;
          left: 50%;
          transform: translate(-50%, -50%);
          transition: 0.25s all ease-in-out;
          border-radius: 50%;
        }

        &.X {
          &::before,
          &::after {
            height: 3px;
            border-radius: 999px;
            background: rgb(0, 188, 100);
            width: 20px;
          }

          &::before {
            transform: translate(-50%, -50%) rotate(45deg);
          }

          &::after {
            transform: translate(-50%, -50%) rotate(135deg);
          }
        }

        &.O {
          &::before {
            border: 3px solid rgb(35, 35, 35);
            width: 12px;
            height: 12px;
          }
        }
      }
    }

    .buttons__wrapper {
      margin-top: 24px;
      display: flex;
      justify-content: center;
    }

    .form {
      margin-bottom: 16px;
      input {
        outline: none;
        box-shadow: none;
        border-radius: 4px;
        border: 1.5px solid #f3f4f7;
        padding: 12px 16px;
      }
    }
  }

  .saved-games {
    width: 320px;
    height: 500px;
    border-radius: 4px;
    border: 1.5px solid #f3f4f7;
    padding: 16px;
    display: flex;
    flex-direction: column;
    align-items: flex-start;

    h3 {
      margin-top: 0;
    }

    p {
      margin: 0;
    }

    .games-list {
      list-style: none;
      height: 400px;
      overflow: hidden auto;
      padding: 0 0;
      margin: 0;
      width: 100%;

      li {
        margin-bottom: 8px;
        display: flex;
        align-items: center;
        justify-content: space-between;
        transition: 0.25s all ease-in-out;
        height: 50px;
        padding: 12px 16px;
        border-radius: 4px;

        .text__wrapper {
          display: flex;
          flex-direction: column;
          align-items: flex-start;

          .timestamp {
            font-size: 12px;
          }
        }

        button {
          display: none;
        }

        &:hover {
          background: #f3f4f7;
          button {
            display: block;
          }
        }
      }
    }
  }
}
