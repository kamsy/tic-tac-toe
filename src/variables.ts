const PLAYER_ONE = "X";
const PLAYER_TWO = "O";
const defaultSize = 3;

export { PLAYER_TWO, PLAYER_ONE, defaultSize };
