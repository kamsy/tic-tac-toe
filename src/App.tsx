import React, { useEffect, useState } from "react";
import "./App.scss";
import { defaultSize, PLAYER_ONE, PLAYER_TWO } from "variables";
import {
  defaultPicks,
  getSavedGamesFromStorage,
  getWinningCombs,
  saveGameToStorage,
  uuid,
} from "utils";
import {
  PicksProps,
  SavedGameProp,
  StringObj,
  WinComboProp,
} from "types/index.types";
import dayjs from "dayjs";
import relativeTime from "dayjs/plugin/relativeTime";
import cx from "classnames";
dayjs.extend(relativeTime);

function App() {
  const [savedGames, setSavedGames] = useState<SavedGameProp[]>([]);
  const [loadedGame, setLoadedGame] = useState<SavedGameProp | null>(null);

  useEffect(() => {
    const gamesFromStorage = getSavedGamesFromStorage();
    setSavedGames(gamesFromStorage);
  }, []);

  const [gridSize, setGridSize] = useState<number>(defaultSize);

  const winningCombos: WinComboProp[] = getWinningCombs(gridSize);
  const [activePlayer, setActivePlayer] = useState<string>(PLAYER_ONE);
  const [playerPicks, setPlayerPicks] = useState<PicksProps>(defaultPicks);
  const [grid, setGrid] = useState<StringObj>({});
  const [winner, setWinner] = useState<null | string>(null);

  const resetGame = () => {
    setPlayerPicks(defaultPicks);
    setGrid({});
    setWinner(null);
    setActivePlayer(PLAYER_ONE);
  };

  const handleSizeUpdate = (e: any) => {
    e.preventDefault();
    const size = +e.target.elements[0].value;
    if (size === gridSize) return;
    if (size < 3) return alert("Grid cannot be less than 3x3");
    resetGame();
    setGridSize(size);
  };

  const handleCellClick = (col: string) => {
    const picked = grid[col];
    if (picked) return;
    setActivePlayer((prev) => (prev === PLAYER_ONE ? PLAYER_TWO : PLAYER_ONE));
    const newPicks = [...playerPicks[activePlayer]];
    newPicks.push(col);
    setPlayerPicks((prev) => ({
      ...prev,
      [activePlayer]: newPicks,
    }));
    setGrid((prev) => ({ ...prev, [col]: activePlayer }));
    winningCombos.forEach((combo) => {
      const isWinner = combo.every((choices) => newPicks.includes(choices));
      if (isWinner) {
        setTimeout(() => {
          setWinner(activePlayer);
        }, 100);
      }
    });
  };

  const saveGame = () => {
    const stem = {
      winner,
      playerPicks,
      grid,
      activePlayer,
      gridSize,
      saved_at: new Date(),
    };
    const gameToSave: SavedGameProp = loadedGame
      ? { ...loadedGame, ...stem }
      : {
          ...stem,
          id: uuid(),
          title: `Game ${savedGames.length + 1}`,
        };

    const _games = [...savedGames];
    const gameIdx = _games.findIndex((game) => game.id === gameToSave.id);
    if (gameIdx !== -1) {
      _games.splice(gameIdx, 1, gameToSave);
      setSavedGames(_games);
      saveGameToStorage(gameToSave);
      return resetGame();
    }
    setSavedGames([gameToSave, ..._games]);
    saveGameToStorage(gameToSave);
    resetGame();
  };

  const loadGame = (id: string) => {
    const gameToLoad = savedGames.find((game) => game.id === id);
    if (!gameToLoad) return;
    const { winner, playerPicks, grid, activePlayer, gridSize } = gameToLoad;
    setLoadedGame(gameToLoad);
    setWinner(winner);
    setPlayerPicks(playerPicks);
    setGrid(grid);
    setActivePlayer(activePlayer);
    setGridSize(gridSize);
  };

  const gridSquare = gridSize ** 2;

  return (
    <div className="App">
      {winner && <h2 className="winner">Player {winner} wins!ðŸŽ‰</h2>}
      <div className="game">
        {loadedGame && <h3>Now playing: {loadedGame.title}</h3>}
        <form onSubmit={handleSizeUpdate} className="form">
          <input placeholder="Grid size" />
          <button>Update</button>
        </form>

        <p className="turn__indicator">Turn {activePlayer}</p>
        <div
          className={cx("grid", { disable: winner })}
          style={{
            gridTemplateColumns: `repeat(${gridSize}, 56px)`,
          }}
        >
          {[...Array.from(Array(gridSquare).keys())].map((key) => {
            const col = key.toString();
            const selected = grid[col];
            return (
              <div
                key={col}
                className={cx("grid-cell", {
                  disable: !!selected,
                })}
                role="button"
                onClick={() => (winner ? null : handleCellClick(col))}
              >
                {selected}
              </div>
            );
          })}
        </div>

        <div className="buttons__wrapper">
          <button onClick={resetGame} className="reset">
            Reset
          </button>
          <button onClick={saveGame}>Save game</button>
        </div>
      </div>

      <div className="saved-games">
        <h3>Saved games</h3>
        <ul className="games-list">
          {!savedGames.length ? (
            <p>No saved games</p>
          ) : (
            savedGames.map(({ title, gridSize, saved_at, id }) => (
              <li key={id}>
                <div className="text__wrapper">
                  <span>
                    {title} ({gridSize}x{gridSize})
                  </span>
                  <span className="timestamp">{dayjs(saved_at).fromNow()}</span>
                </div>

                <button onClick={() => loadGame(id)}>Load</button>
              </li>
            ))
          )}
        </ul>
      </div>
    </div>
  );
}

export default App;
