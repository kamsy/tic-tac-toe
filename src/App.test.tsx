import React from "react";
import { render, screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "./App";
import { PLAYER_ONE, PLAYER_TWO } from "variables";

const setup = () => {
  const view = render(<App />);
  const gridCells = screen.getAllByTestId("grid__cell");
  return { view, gridCells };
};

test("renders correctly", async () => {
  const { view } = setup();
  expect(view).toMatchSnapshot();
  expect(screen.getByTestId("no__games").textContent).toMatchInlineSnapshot(
    `"No saved games"`
  );
  expect(
    screen.getByTestId("turn__indicator").textContent
  ).toMatchInlineSnapshot(`"Turn X"`);
});

test("clicking a box fills it with activeplayer's icon and disables it", async () => {
  const { gridCells } = setup();
  const firstCell = gridCells[0];
  userEvent.click(firstCell);
  expect(firstCell.textContent).toBe(PLAYER_ONE);
  expect(firstCell).toHaveClass("disable");
});

describe("either player can win the game:", () => {
  test(`Player ${PLAYER_ONE} wins:`, async () => {
    const { gridCells } = setup();
    [0, 1, 4, 3, 8, 5].forEach((num) => userEvent.click(gridCells[num]));
    expect(
      (await screen.findByRole("heading", { level: 2 })).textContent
    ).toMatchInlineSnapshot(`"Player X wins!🎉"`);
    expect(screen.getByTestId("grid")).toHaveClass("disable");
  });

  test(`Player ${PLAYER_TWO} wins:`, async () => {
    const { gridCells } = setup();
    [1, 0, 3, 4, 5, 8].forEach((num) => userEvent.click(gridCells[num]));
    expect(
      (await screen.findByRole("heading", { level: 2 })).textContent
    ).toMatchInlineSnapshot(`"Player O wins!🎉"`);
  });
});

test("players are notified when it's a draw", async () => {
  const { gridCells } = setup();
  [0, 1, 2, 6, 7, 8, 3, 4, 5].forEach((num) => userEvent.click(gridCells[num]));
  expect(
    (await screen.findByRole("heading", { level: 2 })).textContent
  ).toMatchInlineSnapshot(`"Game over! Draw🤝"`);
  expect(screen.getByTestId("grid")).toHaveClass("disable");
});

test("can set the grid size to >= 3x3:", async () => {
  render(<App />);
  const gridCells = screen.getAllByTestId("grid__cell");
  expect(gridCells.length).toBe(9);
  const input = screen.getByPlaceholderText(/grid size/i);
  const text = "5";
  userEvent.type(input, text);
  const updateBtn = screen.getByRole("button", {
    name: /Update/i,
  });
  userEvent.click(updateBtn);
  const updatedGridCells = screen.getAllByTestId("grid__cell");
  expect(updatedGridCells.length).toBe((+text) ** 2);
});

test("alerts user grid < 3x3 is set:", async () => {
  render(<App />);
  const alertMock = jest.spyOn(window, "alert").mockImplementation();
  const input = screen.getByPlaceholderText(/grid size/i);
  userEvent.type(input, "2");
  const updateBtn = screen.getByRole("button", {
    name: /Update/i,
  });
  userEvent.click(updateBtn);
  expect(alertMock).toHaveBeenCalledTimes(1);
});

test("saves a game and resets the grid:", async () => {
  const { gridCells } = setup();
  const randomArr = [0, 1, 4, 7];
  randomArr.forEach((num) => userEvent.click(gridCells[num]));
  expect(screen.getByTestId("no__games")).toBeInTheDocument();
  const saveBtn = screen.getByRole("button", {
    name: /Save game/i,
  });
  userEvent.click(saveBtn);
  expect(screen.queryByTestId("no__games")).not.toBeInTheDocument();
  randomArr.forEach((num) => expect(gridCells[num].textContent).toBe(""));
});

test("can load a saved game", async () => {
  const { gridCells } = setup();
  const randomArr1 = [0, 1, 4, 7];
  const randomArr2 = [0, 1, 4, 7];
  const arr = [randomArr1, randomArr2];
  const saveBtn = screen.getByRole("button", {
    name: /Save game/i,
  });
  arr.forEach((arr) => {
    userEvent.click(saveBtn);
    arr.forEach((num) => userEvent.click(gridCells[num]));
  });

  const savedGamesList = screen.getByRole("list");
  expect(savedGamesList.childNodes).toHaveLength(3);
  const gameToLoad = screen.queryAllByTestId("saved__game")[0];
  const loadBtn = within(gameToLoad).getByRole("button", { name: /load/i });
  userEvent.click(loadBtn);
  expect(
    (await screen.findByTestId("loaded__game_title")).textContent
  ).toMatchInlineSnapshot(`"Now playing: Game 3"`);
});
