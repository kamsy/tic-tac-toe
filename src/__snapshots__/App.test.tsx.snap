// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="App"
      >
        <div
          class="game"
        >
          <form
            class="form"
          >
            <input
              placeholder="Grid size"
            />
            <button>
              Update
            </button>
          </form>
          <p
            class="turn__indicator"
            data-testid="turn__indicator"
          >
            Turn 
            X
          </p>
          <div
            class="grid"
            data-testid="grid"
            style="grid-template-columns: repeat(3, 56px);"
          >
            <div
              class="grid-cell"
              data-testid="grid__cell"
              role="button"
            >
              <span />
            </div>
            <div
              class="grid-cell"
              data-testid="grid__cell"
              role="button"
            >
              <span />
            </div>
            <div
              class="grid-cell"
              data-testid="grid__cell"
              role="button"
            >
              <span />
            </div>
            <div
              class="grid-cell"
              data-testid="grid__cell"
              role="button"
            >
              <span />
            </div>
            <div
              class="grid-cell"
              data-testid="grid__cell"
              role="button"
            >
              <span />
            </div>
            <div
              class="grid-cell"
              data-testid="grid__cell"
              role="button"
            >
              <span />
            </div>
            <div
              class="grid-cell"
              data-testid="grid__cell"
              role="button"
            >
              <span />
            </div>
            <div
              class="grid-cell"
              data-testid="grid__cell"
              role="button"
            >
              <span />
            </div>
            <div
              class="grid-cell"
              data-testid="grid__cell"
              role="button"
            >
              <span />
            </div>
          </div>
          <div
            class="buttons__wrapper"
          >
            <button
              class="reset"
            >
              Reset
            </button>
            <button>
              Save game
            </button>
          </div>
        </div>
        <div
          class="saved-games"
        >
          <h3>
            Saved games
          </h3>
          <ul
            class="games-list"
          >
            <p
              data-testid="no__games"
            >
              No saved games
            </p>
          </ul>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="App"
    >
      <div
        class="game"
      >
        <form
          class="form"
        >
          <input
            placeholder="Grid size"
          />
          <button>
            Update
          </button>
        </form>
        <p
          class="turn__indicator"
          data-testid="turn__indicator"
        >
          Turn 
          X
        </p>
        <div
          class="grid"
          data-testid="grid"
          style="grid-template-columns: repeat(3, 56px);"
        >
          <div
            class="grid-cell"
            data-testid="grid__cell"
            role="button"
          >
            <span />
          </div>
          <div
            class="grid-cell"
            data-testid="grid__cell"
            role="button"
          >
            <span />
          </div>
          <div
            class="grid-cell"
            data-testid="grid__cell"
            role="button"
          >
            <span />
          </div>
          <div
            class="grid-cell"
            data-testid="grid__cell"
            role="button"
          >
            <span />
          </div>
          <div
            class="grid-cell"
            data-testid="grid__cell"
            role="button"
          >
            <span />
          </div>
          <div
            class="grid-cell"
            data-testid="grid__cell"
            role="button"
          >
            <span />
          </div>
          <div
            class="grid-cell"
            data-testid="grid__cell"
            role="button"
          >
            <span />
          </div>
          <div
            class="grid-cell"
            data-testid="grid__cell"
            role="button"
          >
            <span />
          </div>
          <div
            class="grid-cell"
            data-testid="grid__cell"
            role="button"
          >
            <span />
          </div>
        </div>
        <div
          class="buttons__wrapper"
        >
          <button
            class="reset"
          >
            Reset
          </button>
          <button>
            Save game
          </button>
        </div>
      </div>
      <div
        class="saved-games"
      >
        <h3>
          Saved games
        </h3>
        <ul
          class="games-list"
        >
          <p
            data-testid="no__games"
          >
            No saved games
          </p>
        </ul>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
